bn_s_mp_toom_sqr.c

```diff
@@ -4,209 +4,177 @@
-   }
-   /* r1 - r2 */
-   if ((res = mp_sub(&w1, &w2, &w1)) != MP_OKAY) {
-      goto LBL_ERR;
-   }
-   /* r3 - r2 */
-   if ((res = mp_sub(&w3, &w2, &w3)) != MP_OKAY) {
-      goto LBL_ERR;
-   }
-   /* r1 - 8r0 */
-   if ((res = mp_mul_2d(&w0, 3, &tmp1)) != MP_OKAY) {
-      goto LBL_ERR;
-   }
-   if ((res = mp_sub(&w1, &tmp1, &w1)) != MP_OKAY) {
-      goto LBL_ERR;
-   }
-   /* r3 - 8r4 */
-   if ((res = mp_mul_2d(&w4, 3, &tmp1)) != MP_OKAY) {
-      goto LBL_ERR;
-   }
-   if ((res = mp_sub(&w3, &tmp1, &w3)) != MP_OKAY) {
-      goto LBL_ERR;
-   }
-   /* 3r2 - r1 - r3 */
-   if ((res = mp_mul_d(&w2, 3uL, &w2)) != MP_OKAY) {
-      goto LBL_ERR;
-   }
-   if ((res = mp_sub(&w2, &w1, &w2)) != MP_OKAY) {
-      goto LBL_ERR;
-   }
-   if ((res = mp_sub(&w2, &w3, &w2)) != MP_OKAY) {
-      goto LBL_ERR;
-   }
-   /* r1 - r2 */
-   if ((res = mp_sub(&w1, &w2, &w1)) != MP_OKAY) {
-      goto LBL_ERR;
-   }
-   /* r3 - r2 */
-   if ((res = mp_sub(&w3, &w2, &w3)) != MP_OKAY) {
-      goto LBL_ERR;
-   }
-   /* r1/3 */
-   if ((res = mp_div_3(&w1, &w1, NULL)) != MP_OKAY) {
-      goto LBL_ERR;
+
+   /** \\S2 = tmp - S4 -S0  */
+   /** S2 = tmp - S4;  */
+   if ((e = mp_sub(&tmp, &S4, &S2)) != MP_OKAY) {
+      goto LTM_ERR;
    }
-   /* r3/3 */
-   if ((res = mp_div_3(&w3, &w3, NULL)) != MP_OKAY) {
-      goto LBL_ERR;
+   /** S2 = S2 - S0;  */
+   if ((e = mp_sub(&S2, &S0, &S2)) != MP_OKAY) {
+      goto LTM_ERR;
    }
 
-   /* at this point shift W[n] by B*n */
-   if ((res = mp_lshd(&w1, 1*B)) != MP_OKAY) {
-      goto LBL_ERR;
+
+   /** P = S4*x^4 + S3*x^3 + S2*x^2 + S1*x + S0; */
+   mp_zero(&tmp);
+   if ((e = mp_lshd(&S4, 4 * B)) != MP_OKAY) {
+      goto LTM_ERR;
    }
-   if ((res = mp_lshd(&w2, 2*B)) != MP_OKAY) {
-      goto LBL_ERR;
+   if ((e = mp_add(&tmp, &S4, &tmp)) != MP_OKAY) {
+      goto LTM_ERR;
    }
-   if ((res = mp_lshd(&w3, 3*B)) != MP_OKAY) {
-      goto LBL_ERR;
+   if ((e = mp_lshd(&S3, 3 * B)) != MP_OKAY) {
+      goto LTM_ERR;
    }
-   if ((res = mp_lshd(&w4, 4*B)) != MP_OKAY) {
-      goto LBL_ERR;
+   if ((e = mp_add(&tmp, &S3, &tmp)) != MP_OKAY) {
+      goto LTM_ERR;
    }
-
-   if ((res = mp_add(&w0, &w1, b)) != MP_OKAY) {
-      goto LBL_ERR;
+   if ((e = mp_lshd(&S2, 2 * B)) != MP_OKAY) {
+      goto LTM_ERR;
+   }
+   if ((e = mp_add(&tmp, &S2, &tmp)) != MP_OKAY) {
+      goto LTM_ERR;
    }
-   if ((res = mp_add(&w2, &w3, &tmp1)) != MP_OKAY) {
-      goto LBL_ERR;
+   if ((e = mp_lshd(&S1, 1 * B)) != MP_OKAY) {
+      goto LTM_ERR;
    }
-   if ((res = mp_add(&w4, &tmp1, &tmp1)) != MP_OKAY) {
-      goto LBL_ERR;
+   if ((e = mp_add(&tmp, &S1, &tmp)) != MP_OKAY) {
+      goto LTM_ERR;
    }
-   if ((res = mp_add(&tmp1, b, b)) != MP_OKAY) {
-      goto LBL_ERR;
+   if ((e = mp_add(&tmp, &S0, &tmp)) != MP_OKAY) {
+      goto LTM_ERR;
    }
+   /** a^2 - P  */
+   mp_exch(&tmp, b);
 
-LBL_ERR:
-   mp_clear_multi(&w0, &w1, &w2, &w3, &w4, &a0, &a1, &a2, &tmp1, NULL);
-   return res;
+LTM_ERR:
```

Comment:
I preferred the old version here for simplicity ;)

Reply:
me too